unified-runtime/source/loader/layers/sanitizer/tsan/tsan_buffer.cpp

```diff
@@ -186,8 +186,10 @@ ur_result_t MemBuffer::getHandle(ur_device_handle_t Device, char *&Handle) {
 }
 
 ur_result_t MemBuffer::free() {
-  for (const auto &[_, Ptr] : Allocations) {
-    ur_result_t URes = getTsanInterceptor()->releaseMemory(Context, Ptr);
+  for (const auto &[Device, Ptr] : Allocations) {
+    ur_result_t URes = Device
+                           ? getTsanInterceptor()->releaseMemory(Context, Ptr)
+                           : getContext()->urDdiTable.USM.pfnFree(Context, Ptr);
```

Comment:
IMO, use urUSMFree in tsan_ddi would be better.

Reply:
The memory is only for internal using, there is no need to call urUSMFree in tsan_ddi